### ISSUE TITLE
OMG YES

### ISSUE THREAD

**Issue by JohnnyWalkerDigital** [zakirullin/cognitive-load#8]
This isn't an "issue" but I just wanted to write a note of support 🎉

A lot of the useful tips and tools mentioned in this document are worth learning and using. OOP, Design Patterns, DDD, small method lengths (side note: never heard anyone say "small classes" -- that one makes no sense to me) are all extremely helpful in the pursuit of easily maintainable/extendable/scalable code... IF they're applied correctly.

I guess the issue is that you might need to apply them incorrectly a few times to learn their benefits.

But as I see it, a general rule of thumb is: Is the thing I'm trying to apply to my code making my FUTURE life harder or easier? When I return to this codebase in six months and have completely forgotten why I made the decisions I did, is the thing I'm trying to apply now actually going to make it easier for me to get back into it?

Or: Is it OK to leave this method long, breaking that recommendation, as long as the method is well-structured, has clearly understandable variables and logic, and has comments to explain what's going on?

You just want to make your future life as easy as possible for yourself (which also makes it easier for your follow team mates).

Rather than trying to be clever, using obscure functions and trying to reduce the number of lines of code, we should be focussing on **reducing the number of lines you need to scroll up in order to understand what's going on**. That's a much better goal than playing code golf with yourself/your team.

Impress me with how BASIC you've made everything (and, when applied correctly, the recommendations mentioned in this article with HELP with that).

**Comment by zakirullin** [comment]
@JohnnyWalkerDigital btw can you advice something to replace "small classes" with? I am not a native speaker unfortunately, and some wording maybe kinda incorrect

**Comment by pixelbrackets** [comment]
:100: :clap: My favorite quote regarding cognitive load in code development is

> Code is a way you treat your coworkers.
> https://twitter.com/mfeathers/status/1031176879577780224

YOU might be able understand a huge code base, and remember links, abbreviations and “historical reasons”. Your colleagues don't. Neither will you, in a year from now or after a job change.

Writing code and designing software architecture that meets complex requirements, but is still readable and easy to understand, is something that will ultimately help both your colleagues and yourself.

**Comment by pixelbrackets** [comment]
Great discussion with answers which exactly show the point: some developers suggest to keep »classes as small as possible« to have glorious »clean code«: https://softwareengineering.stackexchange.com/questions/125357/keep-my-classes-and-methods-as-small-as-possible

**Comment by JohnnyWalkerDigital** [comment]
@pixelbrackets Yes indeed. That illustrates the point very well and I agree with the answers that say "small classes" is a useful and helpful intention expressed poorly. I cracked open my copy of Clean Code to take a look. As you might expect, the book fills the confusion gap very succinctly:

> The first rule of classes is that they should be small. The second rule of classes is that they should be smaller than that. No, we’re not going to repeat the exact same text from the Functions chapter. But as with functions, smaller is the primary rule when it comes to designing classes. As with functions, our immediate question is always “How small?”
> With functions we measured size by counting physical lines. **With classes we use a different measure. We count responsibilities.**

As with all programming best practices, it seems they can be easily misinterpreted!

**Comment by zakirullin** [comment]
We can continue discussion if you don't mind :)

> The idea that one class should have many responsibilities, and that long complex methods are preferable to shorter ones, flies in the face of my experience.
That's not really what I had in mind. Maybe some article modifications are needed 🤔

That's by no means a call to write god objects with many responsibilities

**Comment by JohnnyWalkerDigital** [comment]
Ok, I've just learned more about John Ousterhout's work. I think I understand why I was having so much difficulty [Too many small methods, classes or modules](https://github.com/zakirullin/cognitive-load#too-many-small-methods-classes-or-modules). However, I also agree with Robert Martin's arguments that classes should have single responsibilities.

There must be a middle-path.

Let's see if I can communicate why, or even if what I'm about to say makes any sense...!

As you know, Ousterhout is concerned with the cost of complexity, and he talks about how Deep Modules are preferable to Shallow Ones. He says that a simple interface with complex functionality is more desirable than _vise versa_... and it's hard to disagree!

For him it's about the cost of complexity vs the reward. 

So, if we have to pay 1 mental unit (representing cognitive load) to use a module's interface, then it would be great to get 10+ mental unit's worth of functionality in return! Similarly, if an interface is extremely complicated, but offers little in functionality, then we might pay 10MU to get our heads around the interface in return for 1MU functionality. Not good.

But if we had to pick one out of the two, I'd say that the most important is the _interface_. Yes, taken literally then both are equally important, but in practice **the more complicated the interface, the less likely a net positive outcome will be.**

We don't ever want to be in a net negative mental energy situation where the interface costs more in cognitive load than the functionality we get back. And provided we always make a net positive return on our investment, even if it's 1MU for 2MU return, we're winning. 

Simplifying the interface is generally always going to be much easier than adding functionality to justify the interface usage costs. So we should focus out programmer mental energy there: Writing simple interfaces.

At the same time, I think when Robert Martin wrote, "the first rule of classes is that they should be small. The second rule of classes is that they should be smaller than that" he wasn't talking about the functionality of the classes. He was talking about the **interfaces**.

That is to say: He was trying to simplify a class to the point where the complexity cost for using it was reduced as much as possible. As he wrote in Clean Code:

> The name of a class should describe what responsibilities it fulfills. In fact, naming is probably the first way of helping determine class size. If we cannot derive **a concise name** for a class, then it’s likely too large. The more **ambiguous** the class name, the more likely it has too many responsibilities.

And:

> Every sizable system will contain a large amount of logic and complexity. The primary goal in managing such complexity is to organize it so that a developer knows where to look to find things and **need only understand the directly affected complexity** at any given time. In contrast, a system with larger, multipurpose classes always hampers us by insisting we wade through lots of things we don’t need to know right now.

I would say that Martin is arguing for the same thing as Ousterhout, it's just that people sometimes take Martin's words to an extreme: A class for every _function_ instead of a class for every _responsibility_.

I believe Martin would be perfectly happy a deep class called `blogPost()`, with many methods, provided that class only concerned itself with the status of a blog post. 

What do you think? 🙂 🤔

**Comment by TiffQ** [comment]
I'm a product owner not a developer and even for me this hit the nail on the head. I work on projects where my dev team is constantly changing so overly complex structures make the learning curve for each new team member that much more difficult and the time to usefulness as a team member that much longer.
I am going to share this with the senior mgrs in the hopes of some change of practices. I found this via TDLR newsletter. 
Thank you for taking the time to articulate this.

**Comment by zakirullin** [comment]
We've just added a conclusion, hope you'll like it:

*The intricate and multifaceted nature of cognitive load within the realm of comprehension and problem-solving necessitates a diligent and strategic approach in order to navigate the complexities and optimize mental capacity allocation.*

**With the above statement we have just created an unnecessary cognitive load in your head. Do not do this to your colleagues.**

**Do not code in gibberish.**

**Comment by TiffQ** [comment]
I read that, thought that soberly academic and complex. Then realized that
was a demo. I like it!

**Comment by zakirullin** [comment]
@pixelbrackets I believe that with the conclusion above we conveyed the main idea of your quote as well :) 
> 💯 👏 My favorite quote regarding cognitive load in code development is
> Code is a way you treat your coworkers.

**Comment by zakirullin** [comment]
For starters I have provided an example from John Ousterhout's book

**Comment by turjachaudhuri-opensource** [comment]
I feel developers look at complexity in different ways.

Some developers are okay with complex interfaces, as long as the intent is well-understood, the outcome is clear, and the complexity is justified.

Some developers might want to break down that complex logic into a subset of small modules, with each having more singular focus, but then you need to create a mental map of A calling B, when, etc

I always took these practices as guiding principles, not as enforceable mandates. 
Also, I have an intrinsic faith in the dev community - I feel they mostly want to do the right thing ; some high-level guidance, and nudging is great, but they can mostly formulate next steps from there.

In this case, for example, the overarching message I give to my dev team is 'write code that you can be proud of', and I have seen this resonates a lot with my team, and they mostly come up with good code that ensures overall cognitive load is less.

Open to inputs from others, though.
